File: data_structure/queue/queue.cc
#include "queue.h"

namespace data_structure {
namespace queue {

template <class T>
Queue<T>::Queue()
    : data_(new T[DEFAULT_CAPACITY]), capacity_(DEFAULT_CAPACITY), head_(0),
      tail_(0) {}

template <class T>
Queue<T>::Queue(uint32_t capacity)
    : data_(new T[capacity]), capacity_(capacity), head_(0), tail_(0) {}

template <class T> Queue<T>::~Queue() { delete[] data_; }

template <class T> uint32_t Queue<T>::Size() const {
  return (head_ > tail_) ? capacity_ - head_ + tail_ : tail_ - head_;
}

template <class T> bool Queue<T>::IsEmpty() const { return head_ == tail_; }

template <class T> bool Queue<T>::IsFull() const {
  auto temp = (tail_ + 1 == capacity_) ? 0 : tail_ + 1;
  return head_ == temp;
}

template <class T> T Queue<T>::Head() const {
  if (IsEmpty()) {
    throw std::runtime_error("Queue underflow, can not get head.");
  }
  return data_[head_];
}

template <class T> void Queue<T>::Enqueue(const T &value) {
  if (IsFull()) {
    throw std::runtime_error("Queue overflow, can not enqueue.");
  }
  data_[tail_] = value;
  tail_ = (capacity_ == (tail_ + 1)) ? 0 : tail_ + 1;
}

template <class T> T Queue<T>::Dequeue() {
  if (IsEmpty()) {
    throw std::runtime_error("Queue underflow, can not dequeue.");
  }
  auto removed = data_[head_];
  head_ = (capacity_ == head_ + 1) ? 0 : head_ + 1;
  return removed;
}

} // namespace queue
} // namespace data_structure

int main() {
  std::cout << "TEST QUEUE\n" << std::endl;
  data_structure::queue::Queue<int> queue(20);
  std::cout << "Empty: " << queue.IsEmpty() << std::endl;
  std::cout << "Size: " << queue.Size() << std::endl;
  std::cout << "Enqueue 8" << std::endl;
  queue.Enqueue(8);
  std::cout << "Enqueue 6" << std::endl;
  queue.Enqueue(6);
  std::cout << "Enqueue 26" << std::endl;
  queue.Enqueue(26);
  std::cout << "Enqueue 9" << std::endl;
  queue.Enqueue(9);
  std::cout << "Enqueue 7" << std::endl;
  queue.Enqueue(7);
  std::cout << "Enqueue 22" << std::endl;
  queue.Enqueue(22);
  std::cout << queue << std::endl;
  std::cout << "Empty: " << queue.IsEmpty() << std::endl;
  std::cout << "Size: " << queue.Size() << std::endl;
  std::cout << "Head: " << queue.Head() << std::endl;
  std::cout << "Dequeue: " << queue.Dequeue() << std::endl;
  std::cout << "Head: " << queue.Head() << std::endl;
  std::cout << "Dequeue: " << queue.Dequeue() << std::endl;
  std::cout << "Size: " << queue.Size() << std::endl;
  std::cout << "Dequeue: " << queue.Dequeue() << std::endl;
  std::cout << "Size: " << queue.Size() << std::endl;
  std::cout << queue << std::endl;

  return 0;
}

File: src/data_structure/queue/queue.rs
/// Queue
use std::io::{Error, ErrorKind};

#[derive(PartialEq)]
pub struct Queue<T> {
    data: Vec<T>,
    capacity: usize,
    size: usize,
    front: usize,
    rear: usize,
}

impl<T> Queue<T> {
    pub fn new(capacity: usize) -> Self {
        let mut data = Vec::with_capacity(capacity);
        unsafe {
            data.set_len(capacity);
        }

        Queue {
            data,
            capacity,
            size: 0,
            front: 0,
            rear: capacity - 1,
        }
    }

    pub fn size(&self) -> usize {
        self.size
    }

    pub fn is_empty(&self) -> bool {
        self.size == 0
    }

    pub fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    pub fn front(&self) -> Option<&T> {
        if self.is_empty() {
            return None;
        }

        Some(&self.data[self.front])
    }

    pub fn enqueue(&mut self, data: T) -> Result<(), Error> {
        if self.is_full() {
            return Err(Error::new(
                ErrorKind::Other,
                "Stack overflow, could not push",
            ));
        }

        self.rear = (self.rear + 1) % self.capacity;
        self.data[self.rear] = data;
        self.size += 1;
        Ok(())
    }

    pub fn dequeue(&mut self) -> Result<&T, Error> {
        if self.is_empty() {
            return Err(Error::new(
                ErrorKind::Other,
                "Stack underflow, could not pop",
            ));
        }

        let removed = &self.data[self.front];
        self.front = (self.front + 1) % self.capacity;
        self.size -= 1;
        Ok(removed)
    }
}
